PopPUNK/lineage_clustering.py

```diff
@@ -0,0 +1,680 @@
+                lFile.write(',' + str(lineage_clustering[R][isolate]))
+                lineage_string = str(lineage_clustering[R][isolate])
+                # include information on lineage clustering
+                combined[titles_list[n]][isolate] = lineage_string
+                if lineage_clustering[R][isolate] != previous_lineage_clustering[R][isolate] and previous_lineage_clustering[R][isolate] != null_cluster_value:
+                    lineage_string = str(previous_lineage_clustering[R][isolate]) + ':' + lineage_string
+                if isolate in combined['Overall_lineage'].keys():
+                    combined['Overall_lineage'][isolate] = combined['Overall_lineage'][isolate] + '-' + lineage_string
+                else:
+                    combined['Overall_lineage'][isolate] = lineage_string
+            lFile.write(',' + combined['Overall_lineage'][isolate])
+            if qlist is not None:
+                if isolate in qlist:
+                    lFile.write(',Added')
+                else:
+                    lFile.write(',Existing')
+            lFile.write('\n')
+
+    return combined
+
+def run_clustering_for_R(R, null_cluster_value = None, qlist = None, existing_scheme = False, distances_length = None, distances_type = None):
+    """ Clusters isolates into lineages based on their
+    relative distances using a single R to enable
+    parallelisation.
+
+    Args:
+        R (int)
+            Integer specifying the maximum rank of neighbour used
+            for clustering. Should be changed to int list for hierarchical
+            clustering.
+        null_cluster_value (int)
+            Used to denote as-yet-unclustered isolates.
+        qlist (list)
+            List of query sequences being added to an existing clustering.
+            Should be included within rlist.
+        use_existing (bool)
+            Whether to extend a previously generated analysis or not.
+            
+    Returns:
+        lineage_clustering (dict)
+            Assignment of each isolate to a cluster.
+        lineage_seed (dict)
+            Seed isolate used to initiate each cluster.
+        neighbours (nested dict)
+            Neighbour relationships between isolates for R.
+    """
+    
+    # load shared memory objects
+    existing_distance_shm = shared_memory.SharedMemory(name = 'shm_distances')
+    distances = np.ndarray(distances_length, dtype = distances_type, buffer = existing_distance_shm.buf)
+    existing_rank_shm = shared_memory.SharedMemory(name = 'shm_distance_ranks')
+    distance_ranks = np.ndarray(distances_length, dtype = 'int64', buffer = existing_rank_shm.buf)
+    isolate_list = shared_memory.ShareableList(name = 'shm_isolate_list')
+    
+    # calculate row labels
+    # this is inefficient but there appears to be no way of sharing
+    # strings between processes efficiently
+    row_labels = list(iter(iterDistRows(isolate_list, isolate_list, self = True)))
```

Comment:
This won't scale – 0.5*N^2*2*(~15 char strings)*processes would be ~10Gb for 10k samples, 1Tb for 100k samples

Certainly add the actual labels at the end and use ints to keep track of their index. Also probably better to calculate i,j from row index which mostly avoids this issue

Reply:
distance_ranks is sufficient to do this

Reply:
Functioning update in ed6e729, although some slightly odd output nomenclature.

Reply:
Fixed, and clustering simplified, in ce3a7e9.